srv/apns/push_service.go

```diff
@@ -82,38 +84,81 @@ func (p *pushService) BuildPushServiceProviderFromMap(kv map[string]string, psp
 	}
 
 	if cert, ok := kv["cert"]; ok && len(cert) > 0 {
+		// Binary API
 		psp.FixedData["cert"] = cert
-	} else {
-		return errors.New("NoCertificate")
-	}
 
-	if key, ok := kv["key"]; ok && len(key) > 0 {
-		psp.FixedData["key"] = key
-	} else {
-		return errors.New("NoPrivateKey")
-	}
+		if key, ok := kv["key"]; ok && len(key) > 0 {
+			psp.FixedData["key"] = key
+		} else {
+			return errors.New("NoPrivateKey")
+		}
 
-	_, err := tls.LoadX509KeyPair(psp.FixedData["cert"], psp.FixedData["key"])
-	if err != nil {
-		return err
-	}
+		_, err := tls.LoadX509KeyPair(psp.FixedData["cert"], psp.FixedData["key"])
+		if err != nil {
+			return err
+		}
 
-	if skip, ok := kv["skipverify"]; ok {
-		if skip == "true" {
-			psp.VolatileData["skipverify"] = "true"
+		if skip, ok := kv["skipverify"]; ok {
+			if skip == "true" {
+				psp.VolatileData["skipverify"] = "true"
+			}
 		}
-	}
-	if sandbox, ok := kv["sandbox"]; ok {
-		if sandbox == "true" {
-			psp.VolatileData["addr"] = "gateway.sandbox.push.apple.com:2195"
+
+		if sandbox, ok := kv["sandbox"]; ok {
+			if sandbox == "true" {
+				psp.VolatileData["addr"] = "gateway.sandbox.push.apple.com:2195"
+				return nil
+			}
+		}
+		if addr, ok := kv["addr"]; ok {
+			psp.VolatileData["addr"] = addr
 			return nil
 		}
+		psp.VolatileData["addr"] = "gateway.push.apple.com:2195"
+	} else {
+		if p8, ok := kv["p8"]; ok && len(p8) > 0 {
+			// HTTP API
+			psp.FixedData["p8"] = p8
+
+			_, err := common.LoadPKCS8Key(p8)
+			if err != nil {
+				return err
+			}
+
+			if keyID, ok := kv["keyid"]; ok {
+				psp.FixedData["keyid"] = keyID
+			} else {
+				return errors.New("NoKeyID")
+			}
+
+			if teamID, ok := kv["teamid"]; ok {
+				psp.FixedData["teamid"] = teamID
+			} else {
+				return errors.New("NoTeamID")
+			}
+
+			if bundleID, ok := kv["bundleid"]; ok {
+				psp.FixedData["bundleid"] = bundleID
+			} else {
+				return errors.New("NoBundleID")
+			}
+
+			if sandbox, ok := kv["sandbox"]; ok {
+				if sandbox == "true" {
+					psp.VolatileData["addr"] = "https://api.development.push.apple.com"
+					return nil
+				}
+			}
+			if addr, ok := kv["addr"]; ok {
```

Comment:
I don't see a use case for overriding addr to anything other than those 2 values except unit tests; unit tests are capable of setting the address by other means.